homeassistant/components/enocean/sensor.py

```diff
@@ -46,8 +49,15 @@
         "icon": "mdi:thermometer",
         "class": DEVICE_CLASS_TEMPERATURE,
     },
+    DEVICE_CLASS_WINDOWHANDLE: {
+        "name": "WindowHandle",
+        "unit": "-",
+        "icon": "mdi:window",
+        "class": DEVICE_CLASS_WINDOWHANDLE,
```

Comment:
This device class doesn't exist in home assistant sensor device classes. Just set `None` here.

Reply:
Hi Martin,

I need your help here. Unfortunately, an Enocean device ID doesn't tell you what kind of sensor it is. So in order for the enocean sensor platform to instantiate the correct class, it needs the information about the what kind of sensor it is. See sensor.py, lines 74ff:

```python
def setup_platform(hass, config, add_entities, discovery_info=None):
    """Set up an EnOcean sensor device."""
    dev_id = config.get(CONF_ID)
    dev_name = config.get(CONF_NAME)
    dev_class = config.get(CONF_DEVICE_CLASS)

    if dev_class == DEVICE_CLASS_TEMPERATURE:
        temp_min = config.get(CONF_MIN_TEMP)
        temp_max = config.get(CONF_MAX_TEMP)
        range_from = config.get(CONF_RANGE_FROM)
        range_to = config.get(CONF_RANGE_TO)
        add_entities(
            [
                EnOceanTemperatureSensor(
                    dev_id, dev_name, temp_min, temp_max, range_from, range_to
                )
            ]
        )

    elif dev_class == DEVICE_CLASS_HUMIDITY:
        add_entities([EnOceanHumiditySensor(dev_id, dev_name)])

    elif dev_class == DEVICE_CLASS_POWER:
        add_entities([EnOceanPowerSensor(dev_id, dev_name)])

    elif dev_class == DEVICE_CLASS_WINDOWHANDLE:
        add_entities([EnOceanWindowHandle(dev_id, dev_name)])

```
Right now, if the device class is set to None, no sensor will be instantiated. On the other hand, if the device class is None, defaulting to a window handle sensor is kind of counter intuitive.

How do we solve this?

Reply:
The powersensor already used the same technique for the same reason. It's proven to work.

This could also be solved by adding an extra field for the enocean-device-class, I guess. This would have to be done for all enocean device types, which is not the core of this change, I would say.

To be honest, for a field named `class` to be limited to a predefined set of possible sensor types feels like a ridiculous limitation for a system like homeassistant?

(yes, the enocean integration need refactoring, e.g. the push button totally against HA patterns);

Reply:
The `"class"` key is only used here:
https://github.com/home-assistant/home-assistant/blob/876195a8a8c2e3fed27df0ffa88ef2206e272e56/homeassistant/components/enocean/sensor.py#L97

It's perfectly fine to set its value to `None`. I think you're mixing up the device type with the device class. The device type, the keys in this constant dict is Enocean specific. The device class is Home Assistant specific. To make this more clear, some variables should probably be renamed.

Reply:
The powersensor should be fixed.

Reply:
@MartinHjelmare Ah I think I understood - yes, the naming is confusing. The type in the `SENSOR_TYPES` should be `None`, but using the custom sensor type as configuration value for `device_class` (and used in `setup_platform`) is fine?

Reply:
Yes. We should rename the  `device_class` variable in `setup_platform` to `sensor_type` or similar to distinguish between that and the `device_class` entity property.

Reply:
Ok guys, here is what I have understood:
1. From now on we should speak of sensor types, not of device classes. That means I will rename the two constants in
https://github.com/rhadamantys/home-assistant/blob/069c69e84d371df77f190e253e55f3b4a82cfd52/homeassistant/components/enocean/sensor.py#L30-L31
from `DEVICE_CLASS_*` to `SENSOR_TYPE_*`
2. I will add two more platform specific constants at the same place `SENSOR_TYPE_HUMIDITY` and `SENSOR_TYPE_TEMPERATURE` to become independent from the device class definitions in const.py
3. In the following `SENSOR_TYPES` dictionary, I will replace the dictionary keys with the new `SENSOR_TYPES_*` constants
4. In `setup_platform`, I will rename the local variable `dev_class` with `sensor_type`. That makes the code consistent with the constructors of the individual sensors. There the local variable is already `sensor_type`.
5. That brings me to my last point:
https://github.com/rhadamantys/home-assistant/blob/069c69e84d371df77f190e253e55f3b4a82cfd52/homeassistant/components/enocean/sensor.py#L109-L110
The base class `EnOceanDensor` has two members with identical values: `_sensor_type` and `_device_class`. 
https://github.com/rhadamantys/home-assistant/blob/069c69e84d371df77f190e253e55f3b4a82cfd52/homeassistant/components/enocean/sensor.py#L109-L110
As the definitions in the `SENSOR_TYPE` dictionary above have a circular reference ("sort of": the key value is identical to the sensor type dictionary's `class` value), I would suggest to completely remove the `class` key value pair from the dictionary and consistently use the `_sensor_type` member wherever `_device_class` is used (basically here)
https://github.com/rhadamantys/home-assistant/blob/069c69e84d371df77f190e253e55f3b4a82cfd52/homeassistant/components/enocean/sensor.py#L128-L131
Is that ok? Or do I miss the bigger picture here? 

I hope with this, we can resolve that issue.

Reply:
5. No. `device_class` is a home assistant api and must be kept. The point is that this value is not the same as sensor_type, even though it might overlap in some cases.

1-4 is good!

Reply:
done